### DISCUSSION TITLE
"Safer" single-track train station code - stop/pause as an "if"?

### DISCUSSION THREAD

**Discussion by Rogue75** [pybricks/support#1335] [Q&A]
For context, I am making a single-track railroad line, but it extends very perilously in my home (if the train fails, I need to get the neighbors 20ft ladder...). Therefore, I would like the software to have some "safety" built-in to prevent it going off each end. I will create mechanical ramps at each end too just in case---to lift the wheels up if the color scanner fails and the train starts to run off the track so that it too doesn't fall off a further drop of 30ft...

I'm basing my code off of [Eggybricks example](https://github.com/eggybricks/poweredup-color-sensor/blob/main/pybricks-code-for-60337-express-passenger-train.py). Their track set up is red at one end, yellow at the other end, and green in the middle for the train station. However, their code searches for green _before_ seeing the end (yellow or red). If it misses green, it doesn't stop at yellow or red and [runs right off...](https://media.giphy.com/media/v1.Y2lkPTc5MGI3NjExdm44a2g5NGFwMW96dGt6MThseDBjbXR5bzBqMGszbm05ZTBveHl1MSZlcD12MV9pbnRlcm5hbF9naWZfYnlfaWQmY3Q9Zw/JYEEdjzlsb280/giphy.gif)

### To clarify...
I want the train to always look for each end, and _if_ it sees the train station color to pause, before proceeding in the same direction it was headed.

Here's my attempt at that functionality and code (I'm new to Python):

```python
def check_for_color(color):
    while sensor.color() != color:
        wait(check_color_interval_ms)

while True:  
    print("looking for yellow or green")
    check_for_color(Color.YELLOW or Color.GREEN)

if color = Color.YELLOW
    print("yellow detected, at end of line, stopping and going back toward station")
    print(sensor.hsv())
    motor.brake()
    wait(eol_stop_time_ms)
    motor.dc(forward_speed)

if color = Color.GREEN
    print("green detected, at station, stopping and continuing")
    print(sensor.hsv())
    motor.brake()
    wait(station_stop_time_ms)
    motor.dc(-forward_speed)

    then 
            print("looking for yellow")
            check_for_color(Color.YELLOW)

            print("yellow detected, at end of line, stopping and going back to station")
            print(sensor.hsv())
            motor.brake()
            wait(eol_stop_time_ms)
            motor.dc(-forward_speed)

    print("looking for red or green")
    check_for_color(Color.RED or Color.GREEN)

if color = Color.RED
    print("red detected, at end of line, stopping and going back toward station")
    print(sensor.hsv())
    motor.brake()
    wait(eol_stop_time_ms)
    motor.dc(-forward_speed)

if color = Color.GREEN
    print("green detected, at station, stopping and continuing")
    print(sensor.hsv())
    motor.brake()
    wait(station_stop_time_ms)
    motor.dc(forward_speed)

    then 
            print("looking for red")
            check_for_color(Color.RED)

            print("red detected, at end of line, stopping and going back toward station")
            print(sensor.hsv())
            motor.brake()
            wait(eol_stop_time_ms)
            motor.dc(-forward_speed)

   
```

After writing this up, I read some of the comments on the [Eggybricks youtube video](https://www.youtube.com/watch?v=Jmh4NLxJ_8c) and some folks had some suggestions to do something like this...

1 commenter
>  As a suggestion for V2, I'd say put the "current speed" into a variable, and have your colour detection be a switch statement. Then the "Green" case can stop and wait, then re-apply the speed from the variable, the "Yellow" case stops, waits, changes the speed variable to - speed, and sets that speed, and the "Red" case does the same, but with + speed. Then it'll work for tracks with more than one station, and rather than having to push the train to the station when it starts, it can check the sensor and if nothing's detected, start going forward. 

2nd commenter
> Really like this setup, but I have issues with your code. It is assuming that the train is between the red and yellow at startup. If it is not, it will run past the yellow and off the track. I would suggest you change your check_for_color to not have a parameter and to have a switch statement to see if it has one of the three command colors. Then depending on what color it sees it can stop or change directions. You will have to add a direction variable to allow the program to remember what direction it was going in the station and continue in that direction after the hold time. In fact, this would allow you to have just two colors (red and green for example). When it hits red it just looks to see what direction it is going, then change to the other.
> 
> With this setup, the program will always stop at a reverse point and will not be dependent on a starting position.

3rd commenter
> The PU Sensor is not as good as the Mindstorms/Education sensor. The light with it seems to increase the accuracy. I also coded something similar a while back before I saw your video, and I managed to make it work using just two colours. I just stored a variable for the speed of the motor and modified that every time I saw the colour. You could do more stuff with other colours then if you had more ports available! 

I'm not sure my example works (I haven't tested it, but I doubt it as some of the code isn't colored correctly in my editor) or how to go about doing a switch statement or what the best path is to accomplish this. Thanks in advance for any guidance.

**Comment by laurensvalk** [comment]
```python
def check_for_color(color):
    while sensor.color() != color:
        wait(check_color_interval_ms)

while True:  
    print("looking for yellow or green")
    check_for_color(Color.YELLOW or Color.GREEN)

# rest of the code will not run
```

This `while` loop will keep running forever, so the rest of the code below it will not run.

> I'm new to Python

You might like to check out the block version of Pybricks. It so happens we just made a tutorial for your project :smile: :

You can see the generated Python code as you go, to see it all come together. This will also let you compare it with your experiments to made far.

[![Train video](https://img.youtube.com/vi/kwnSbnlhOkA/0.jpg)](https://www.youtube.com/watch?v=kwnSbnlhOkA)

**Comment by Rogue75** [comment]
Thanks! It's more that I wanted to learn how to code some python while doing this, so thank you for pointing out that I can save the code and read it. It's awesome that it also includes comments in the code so I can better wrap my head around what's happening.

I'm giving the block version a try now. During this, I confirmed what you guys mentioned in my other thread. The colors are indeed dark. Do you think it makes sense to rebuild the train to try and get the sensor lower to get brighter colors?

<img width="397" alt="colors dark" src="https://github.com/pybricks/support/assets/7265814/b4c0b320-3755-4faf-8bf7-c4c7ac555f65">

**Reply by laurensvalk** [reply]
Thanks for signing up, we appreciate it :smile: 

> During this, I confirmed what you guys mentioned in my other thread. The colors are indeed dark.

That's a nice visual indeed.

> Do you think it makes sense to rebuild the train to try and get the sensor lower to get brighter colors?

I would say so. Changing software can only do so much, especially if physics and sensor quality are in the way.

**Comment by Rogue75** [comment]
You guys are rad, I just got the lifetime membership!

Update: After a heavy rebuild, it's much closer to the tracks now and the colors look good!
<img width="573" alt="new colors" src="https://github.com/pybricks/support/assets/7265814/67354989-abcf-43d0-b1dc-d01ff980ee58">
![20231216_213815](https://github.com/pybricks/support/assets/7265814/89b832bc-d8e9-40d7-a961-537125e66e35)

However, my new code doesn't work haha. I think I might be better at brick building than brick coding haha.
I'm used to html, css, and tweaking javascript, so maybe I'm missing some fundamental learnings in logic flow.

<img width="328" alt="update1" src="https://github.com/pybricks/support/assets/7265814/6ed17efe-f4ed-4122-a5d5-809192bac962">

Another attempt which makes sense in my head, but doesn't work as it doesn't stop at the green station.

<img width="359" alt="update2" src="https://github.com/pybricks/support/assets/7265814/a947b956-2df8-4b26-8735-d5cf0a16cf73">

I think I have two challenges here:
1) How do I have the middle green brick be what it's searching for first and then red or yellow comes next so that if it misses green for some reason it can still end at red or yellow?
2) How do I have the green brick train station work both ways? The current code means I have to give a direction to the motor after it sees and pauses at green. So somehow I need a directional trigger?

**Reply by laurensvalk** [reply]
> You guys are rad, I just got the lifetime membership!

Thanks! Don't forget to cancel your monthly trial - unless you like us that much :smile: 

> Update: After a heavy rebuild, it's much closer to the tracks now and the colors look good!

Awesome!

> However, my new code doesn't work

I think there are two distinct things to look at.

**Which colors to distinguish**

In your snippet, the sensor is only set up to distinguish the color markers, but not the track and background. Here's the setup from the tutorial for comparison. The video goes into a bit more detail on how this works.

![image](https://github.com/pybricks/support/assets/12326241/3950475d-66f3-4d7a-a451-53f64011fc64)

**Program logic**

Perhaps you can start with a smaller, separate program to experiment with. Have a look at how loops and if/else switches work.

You can also use the `print` block (`output` category) to print out your own statements like "green detected!" so you can follow along and push the train around manually.

----

Let me know how it goes after these two suggestions :)

**Reply by BertLindeman** [reply]
Laurens, did you paste an image here in stead of a movie? 
I can not get it to "roll" Firefox /Chrome on win11

**Reply by laurensvalk** [reply]
I did, to show that particular view with the colors and the train in one.

Here’s the full video: https://youtu.be/kwnSbnlhOkA

**Reply by Rogue75** [reply]
Thanks, Laurensvalk for the reminder to update my subscription. I did that.

Regarding the "Which colors to distinguish", I had a total D'oh! moment and followed along with your tutorial, but somehow missed that color part. That helped.

Regarding the "Program logic", I ran out of time to make it work for the Christmas trolley so I removed the middle train station and just went with one stop on each end. It'll work well for this year, and I can revisit it next time. Thanks for your help! I'll post a video once I'm done with that.
